### PR TITLE
Validate ResourceTemplate params are declared

### LINKED ISSUE #255: TriggerTemplate params work without the spec.params field
# Expected Behavior
I expect that when my TriggerTemplate does not have the `spec.params` field, the resource template parameter substitution will not work.

# Actual Behavior
When I do not specify the `spec.params` field, parameter substitution still appears to work. [Here](https://github.com/pipeline-hotel/example-pipelines/blob/master/triggers-resources/config/simple-pipeline/simple-pipeline-template.yaml) is an example of a TriggerTemplate that works without its `spec.params` field.

# Steps to Reproduce the Problem

1. Delete all the parameters from one of your TriggerTemplates (delete the entire `spec.params` field)
2. Use the TriggerTemplate like you normally would

### PR DESCRIPTION
**Pull request by ncskier**
# Changes

Fixes #255
Adds validation to the TriggerTemplate resource. Validation will fail
when a ResourceTemplate uses a parameter not already declared in
`spec.params`.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/triggers/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

```
TriggerTemplates are now invalid if any of their ResourceTemplates
uses a parameter not declared in spec.params.
```

### CODE REVIEW

File: pkg/apis/triggers/v1alpha1/trigger_template_validation.go [hunk 1]
```diff
@@ -64,3 +69,29 @@ func validateResourceTemplates(templates []TriggerResourceTemplate) *apis.FieldE
...
+
+// Verify every param in the ResourceTemplates is declared with a ParamSpec
+func verifyParamDeclarations(params []pipelinev1.ParamSpec, templates []TriggerResourceTemplate) *apis.FieldError {
+	declaredParamNames := map[string]struct{}{}
+	for _, param := range params {
+		declaredParamNames[param.Name] = struct{}{}
+	}
+	for i, template := range templates {
+		// Get all params in the template $(params.NAME)
+		paramsRegexp := regexp.MustCompile(`\$\(params.(?P<var>[_a-zA-Z][_a-zA-Z0-9.-]*)\)`)
```
Comment (dibyom) [inline pkg/apis/triggers/v1alpha1/trigger_template_validation.go]: Can this regex declaration be moved up outside the for loop?

Comment (ncskier) [reply pkg/apis/triggers/v1alpha1/trigger_template_validation.go]: Definitely, good catch! 🙂

**Comment by ncskier** [comment]
Thanks for the review @dibyom! Sorry it took me so long to respond to your feedback, I just got back from vacation 😄

**Comment by dibyom** [comment]
@ncskier no worries, just getting back today as well :smile:

File: pkg/apis/triggers/v1alpha1/trigger_template_validation.go [hunk 2]
```diff
@@ -64,3 +69,29 @@ func validateResourceTemplates(templates []TriggerResourceTemplate) *apis.FieldE
...
 	return nil
 }
+
+// Verify every param in the ResourceTemplates is declared with a ParamSpec
+func verifyParamDeclarations(params []pipelinev1.ParamSpec, templates []TriggerResourceTemplate) *apis.FieldError {
+	declaredParamNames := map[string]struct{}{}
+	for _, param := range params {
+		declaredParamNames[param.Name] = struct{}{}
+	}
+	paramsRegexp := regexp.MustCompile(`\$\(params.(?P<var>[_a-zA-Z][_a-zA-Z0-9.-]*)\)`)
```
Comment (dibyom) [inline pkg/apis/triggers/v1alpha1/trigger_template_validation.go]: Now that I think about it, can we move the declaration out of the function body itself, maybe to a var/const block in the package?

Comment (ncskier) [reply pkg/apis/triggers/v1alpha1/trigger_template_validation.go]: Yep, that works for me too. Just made the change @dibyom 👍

**Pull request merged**